src/main/java/com/avail/optimizer/L2Optimizer.java

```diff
@@ -465,67 +645,170 @@ void postponeConditionallyUsedValues ()
+		final Set<L2Register> registersConsumedLaterInBlock,
+		final Set<L2SemanticValue> semanticValuesConsumedLaterInBlock,
+		final int readMaskLaterInBlock,
+		final int writeMaskLaterInBlock,
+		final Set<L2PcOperand> candidateTargetEdges)
 	{
-		if (instruction.hasSideEffect()
-			|| instruction.altersControlFlow()
+		if (instruction.altersControlFlow()
 			|| instruction.operation().isPhi()
-			|| instruction.operation().isEntryPoint(instruction))
+			|| instruction.isEntryPoint())
+		{
+			return null;
+		}
+		// TODO Until we've eliminated this flag...
+		if (instruction.hasSideEffect())
 		{
 			return null;
 		}
+		final int writeMask = instruction.operation().writesHiddenVariablesMask;
+		if ((writeMask & readMaskLaterInBlock) != 0
+			|| (writeMask & writeMaskLaterInBlock) != 0)
+		{
+			// Either the instruction writes something that a later instruction
+			// reads, or there are multiple writes that should maintain their
+			// order in case a later block has a dependent read.  Don't move it.
+			return null;
+		}
+		final int readMask = instruction.operation().readsHiddenVariablesMask;
+		if ((readMask & writeMaskLaterInBlock) != 0)
+		{
+			// The instruction reads something that a later instruction would
+			// clobber.  Don't move it.
+			return null;
+		}
+
 		final List<L2Register> written = instruction.destinationRegisters();
+		final List<L2SemanticValue> writtenSemanticValues = new ArrayList<>();
+		instruction.writeOperands().forEach(
+			write -> writtenSemanticValues.addAll(write.semanticValues()));
 		assert !written.isEmpty()
 			: "Every instruction should either have side effects or write to "
 			+ "at least one register";
-		if (!disjoint(written, registersConsumedLaterInSameBlock))
+		if (!disjoint(written, registersConsumedLaterInBlock))
 		{
 			// A later instruction in the current basic block consumes one of
-			// the values produced by the given instruction, so we can't move
+			// the registers produced by the given instruction, so we can't move
 			// the given instruction into later blocks.
 			return null;
 		}
-		final L2BasicBlock block = instruction.basicBlock;
-		if (block.successorEdgesCount() == 1)
+		if (!disjoint(
+			writtenSemanticValues, semanticValuesConsumedLaterInBlock))
+		{
+			// A later instruction in the current basic block consumes one of
+			// the semantic values produced by the given instruction, so we
+			// can't move the given instruction into later blocks.
+			return null;
+		}
+		if (candidateTargetEdges.size() == 1)
 		{
 			// There's only one successor edge.  Since the CFG is in edge-split
 			// form, the successor might have multiple predecessors.  Don't move
 			// across the edge in that case, since it may cause the instruction
 			// to run in situations that it doesn't need to.  When code
 			// splitting is eventually implemented, it should clean up this case
 			// by duplicating the successor block just for this edge.
```

Comment:
There should probably be a `TODO` here since it represents future functionality, and this comment should be cleaned up when code splitting is operational.

Reply:
Split comment, and added TODO to the appropriate half.